File: careerfields/engine.py
"""
Recommendation formula:
  Final = Subject 40% + Interest 35% + Education 15% + Market 10%

Marks are stored as {subject: {obtained, total, percent}}.
Interest uses tag frequency so repeated signals (activities + skill + area) weigh more.
Stream mismatch (e.g. Pre-Medical vs Engineering) is penalized so rankings stay honest.
"""
from collections import Counter

LEVEL_READY = {
    'A-Level': 100,
    'Intermediate': 98,
    'O-Level': 74,
    'Matric': 70,
}

RELATED_STREAMS = {
    'Pre-Engineering': {'ICS': 78, 'Pre-Medical': 28, 'Commerce': 22},
    'ICS': {'Pre-Engineering': 80, 'Commerce': 42, 'Pre-Medical': 24},
    'Pre-Medical': {'Pre-Engineering': 30, 'ICS': 24},
    'Commerce': {'ICS': 48, 'Arts / Humanities': 55},
    'Arts / Humanities': {'Commerce': 52},
}

CATEGORY_TAGS = {
    'Computer Science': {'cs-it-ai', 'computers', 'technology'},
    'Engineering': {'engineering', 'physics', 'math'},
    'Medical': {'medicine', 'biology', 'helping-people'},
    'Business': {'business', 'commerce', 'entrepreneur', 'leading'},
    'Social Sciences': {'social-sciences', 'helping-people', 'languages'},
}

SUBJECT_TAGS = {
    'math': 'Mathematics',
    'physics': 'Physics',
    'chemistry': 'Chemistry',
    'biology': 'Biology',
    'computers': 'Computer Science',
    'commerce': 'Accounting',
    'languages': 'English',
}


def mark_percent(entry):
    if entry is None or entry == '':
        return None
    if isinstance(entry, dict):
        if entry.get('percent') is not None and entry.get('percent') != '':
            try:
                return max(0.0, min(100.0, float(entry['percent'])))
            except (TypeError, ValueError):
                pass
        total = entry.get('total')
        obtained = entry.get('obtained')
        try:
            total = float(total)
            obtained = float(obtained)
        except (TypeError, ValueError):
            return None
        if total <= 0:
            return None
        return max(0.0, min(100.0, 100.0 * obtained / total))
    try:
        return max(0.0, min(100.0, float(entry)))
    except (TypeError, ValueError):
        return None


def subject_match_score(field, marks, user_tags=None):
    marks = marks or {}
    required = field.required_subjects or []
    percents = []
    missing = 0
    for sub in required:
        value = mark_percent(marks.get(sub))
        if value is None:
            missing += 1
        else:
            percents.append(value)
    if not required:
        all_p = [mark_percent(v) for v in marks.values()]
        all_p = [p for p in all_p if p is not None]
        return round(sum(all_p) / len(all_p)) if all_p else 50, []

    enjoyed = 0
    user_tags = user_tags or {}
    for tag, subject in SUBJECT_TAGS.items():
        if subject in required and user_tags.get(tag):
            enjoyed += 1

    if not percents:
        return 30, []

    avg = sum(percents) / len(percents)
    score = avg - missing * 14
    score += min(8, enjoyed * 4)
    score = max(0, min(100, score))
    details = [f'{sub} {round(mark_percent(marks.get(sub)) or 0)}%' for sub in required if mark_percent(marks.get(sub)) is not None]
    return round(score), details


def interest_match_score(field, user_tag_weights):
    field_tags = [t for t in (field.interest_tags or []) if t]
    weights = user_tag_weights or {}
    if not field_tags:
        return 50, []
    if not weights:
        return 32, []

    hit = 0.0
    matched = []
    for tag in field_tags:
        w = weights.get(tag, 0)
        if w:
            hit += min(1.0, 0.52 + 0.16 * w)
            matched.append(tag)
    raw = 100.0 * hit / len(field_tags)

    category_boost = 0
    for tag in CATEGORY_TAGS.get(field.category, set()):
        if weights.get(tag):
            category_boost = 8
            break
    raw = max(0, min(100, raw + category_boost))
    return round(raw), matched[:5]


def education_level_score(level):
    if not level:
        return 50
    return LEVEL_READY.get(level, 50)


def background_match_score(field, background):
    allowed = [item for item in (field.min_background or []) if item and item != 'Any']
    if not field.min_background or 'Any' in (field.min_background or []):
        return 100 if background else 50
    if not background:
        return 50
    if background in allowed:
        return 100
    related = RELATED_STREAMS.get(background, {})
    best = 18
    for stream in allowed:
        best = max(best, related.get(stream, 18))
    return best


def education_match_score(field, level, background=''):
    level_part = education_level_score(level)
    stream_part = background_match_score(field, background)
    return round(level_part * 0.45 + stream_part * 0.55)


def market_score(field):
    avg = ((field.market or 5) + (field.future or 5)) / 2
    return round(avg * 10)


def stream_penalty(field, background):
    allowed = field.min_background or []
    if not background or not allowed or 'Any' in allowed:
        return 1.0
    if background in allowed:
        return 1.0
    if field.category == 'Medical':
        return 0.58
    if field.category == 'Engineering' and background not in ('Pre-Engineering', 'ICS'):
        return 0.72
    return 0.88


def score_field(field, marks, user_tag_weights, level, background=''):
    subject, subject_bits = subject_match_score(field, marks, user_tag_weights)
    interest, matched_tags = interest_match_score(field, user_tag_weights)
    education = education_match_score(field, level, background)
    market = market_score(field)
    final = subject * 0.40 + interest * 0.35 + education * 0.15 + market * 0.10
    final *= stream_penalty(field, background)
    final = round(max(0, min(99, final)))

    reasons = []
    if subject_bits:
        reasons.append('Marks: ' + ', '.join(subject_bits[:3]))
    if matched_tags:
        reasons.append('Interests: ' + ', '.join(matched_tags[:4]))
    allowed = field.min_background or []
    if background and (background in allowed or 'Any' in allowed):
        reasons.append(f'Stream fit: {background}')
    elif background:
        reasons.append(f'Stream {background} is a weaker fit for this field')
    if level:
        reasons.append(f'Education level: {level}')
    return {
        'subject': subject,
        'interest': interest,
        'education': education,
        'market': market,
        'final': final,
        'reasons': reasons[:4],
    }


def user_tag_weights(user):
    counts = Counter()
    for answer in user.questionnaire_answers.select_related('option'):
        tag = (answer.option.tag or '').strip()
        if tag:
            counts[tag] += 1
    return dict(counts)


def user_interest_tags(user):
    return list(user_tag_weights(user))

File: frontend/js/engine.js
function markPercent(entry) {
  if (entry == null || entry === '') return null;
  if (typeof entry === 'object') {
    if (entry.percent != null && entry.percent !== '') return Number(entry.percent);
    const total = Number(entry.total);
    const obtained = Number(entry.obtained);
    if (!total) return null;
    return Math.max(0, Math.min(100, 100 * obtained / total));
  }
  const n = Number(entry);
  return Number.isNaN(n) ? null : Math.max(0, Math.min(100, n));
}

function getInterestTagWeights() {
  const weights = {};
  Object.values(state.quizAnswers).forEach((ans) => {
    const tags = Array.isArray(ans) ? ans : [ans];
    tags.forEach((t) => {
      if (!t) return;
      weights[t] = (weights[t] || 0) + 1;
    });
  });
  return weights;
}

function getInterestTags() {
  return Object.keys(getInterestTagWeights());
}

function subjectMatchScore(field) {
  const marks = state.academic.marks || {};
  const required = field.requiredSubjects || [];
  const weights = getInterestTagWeights();
  const SUBJECT_TAGS = {
    math: 'Mathematics', physics: 'Physics', chemistry: 'Chemistry', biology: 'Biology',
    computers: 'Computer Science', commerce: 'Accounting', languages: 'English',
  };
  if (!required.length) {
    const all = Object.values(marks).map(markPercent).filter((p) => p != null);
    return all.length ? Math.round(all.reduce((a, b) => a + b, 0) / all.length) : 50;
  }
  const percents = [];
  let missing = 0;
  required.forEach((sub) => {
    const p = markPercent(marks[sub]);
    if (p == null) missing += 1;
    else percents.push(p);
  });
  if (!percents.length) return 30;
  let enjoyed = 0;
  Object.entries(SUBJECT_TAGS).forEach(([tag, subject]) => {
    if (required.includes(subject) && weights[tag]) enjoyed += 1;
  });
  return Math.round(Math.max(0, Math.min(100, (percents.reduce((a, b) => a + b, 0) / percents.length) - missing * 14 + Math.min(8, enjoyed * 4))));
}

function interestMatchScore(field) {
  const weights = getInterestTagWeights();
  const fieldTags = (field.interestTags || []).filter(Boolean);
  if (!fieldTags.length) return 50;
  if (!Object.keys(weights).length) return 32;
  let hit = 0;
  fieldTags.forEach((tag) => {
    const w = weights[tag] || 0;
    if (w) hit += Math.min(1, 0.52 + 0.16 * w);
  });
  let raw = 100 * hit / fieldTags.length;
  const CATEGORY_TAGS = {
    'Computer Science': ['cs-it-ai', 'computers', 'technology'],
    'Engineering': ['engineering', 'physics', 'math'],
    'Medical': ['medicine', 'biology', 'helping-people'],
    'Business': ['business', 'commerce', 'entrepreneur', 'leading'],
    'Social Sciences': ['social-sciences', 'helping-people', 'languages'],
  };
  if ((CATEGORY_TAGS[field.category] || []).some((t) => weights[t])) raw += 8;
  return Math.round(Math.max(0, Math.min(100, raw)));
}

function educationMatchScore(field) {
  const LEVEL_READY = { 'A-Level': 100, Intermediate: 98, 'O-Level': 74, Matric: 70 };
  const levelPart = LEVEL_READY[state.academic.level] || 50;
  const allowed = (field.minBackground || []).filter((x) => x && x !== 'Any');
  const background = state.academic.background;
  let streamPart = 50;
  if (!(field.minBackground || []).length || (field.minBackground || []).includes('Any')) {
    streamPart = background ? 100 : 50;
  } else if (allowed.includes(background)) {
    streamPart = 100;
  } else if (!background) {
    streamPart = 50;
  } else {
    const RELATED = {
      'Pre-Engineering': { ICS: 78, 'Pre-Medical': 28, Commerce: 22 },
      ICS: { 'Pre-Engineering': 80, Commerce: 42, 'Pre-Medical': 24 },
      'Pre-Medical': { 'Pre-Engineering': 30, ICS: 24 },
      Commerce: { ICS: 48, 'Arts / Humanities': 55 },
      'Arts / Humanities': { Commerce: 52 },
    };
    streamPart = Math.max(18, ...allowed.map((s) => (RELATED[background] || {})[s] || 18));
  }
  return Math.round(levelPart * 0.45 + streamPart * 0.55);
}

function marketScore(field) {
  const avg = ((field.market || 5) + (field.future || 5)) / 2;
  return Math.round(avg * 10);
}

function streamPenalty(field) {
  const background = state.academic.background;
  const allowed = field.minBackground || [];
  if (!background || !allowed.length || allowed.includes('Any') || allowed.includes(background)) return 1;
  if (field.category === 'Medical') return 0.58;
  if (field.category === 'Engineering' && background !== 'Pre-Engineering' && background !== 'ICS') return 0.72;
  return 0.88;
}

function scoreField(field) {
  const subject = subjectMatchScore(field);
  const interest = interestMatchScore(field);
  const education = educationMatchScore(field);
  const market = marketScore(field);
  let final = subject * 0.40 + interest * 0.35 + education * 0.15 + market * 0.10;
  final *= streamPenalty(field);
  final = Math.round(Math.max(0, Math.min(99, final)));
  return {
    ...field,
    scores: { subject, interest, education, market, final },
    match: final,
  };
}

function generateRecommendations() {
  if (state.apiRecs && state.apiRecs.length) return state.apiRecs;
  return FIELDS.map(scoreField).sort((a, b) => b.match - a.match);
}

function profileCompletion() {
  let n = 0;
  if (state.academic.level) n += 25;
  if (state.academic.background) n += 15;
  if (Object.keys(state.academic.marks).length >= 2) n += 20;
  if (state.quizComplete) n += 40;
  return n;
}

function recommendedUnis(topFieldIds) {
  return UNIS.map((u) => {
    const overlap = (u.fieldIds || []).filter((id) => topFieldIds.includes(id)).length;
    return { ...u, relevance: overlap };
  }).filter((u) => u.relevance > 0).sort((a, b) => b.relevance - a.relevance);
}

function recommendedSchols(avg, topCategories) {
  return SCHOLS.filter((s) => {
    if (s.minMarks && avg && avg < s.minMarks) return false;
    const cats = s.fieldCategories || [];
    if (!cats.length || cats.includes('All')) return true;
    const blob = `${s.fieldOfStudy} ${cats.join(' ')}`.toLowerCase();
    return topCategories.some((c) => {
      const key = (c || '').toLowerCase();
      if (blob.includes(key)) return true;
      if (key.includes('computer') && /it|ai|computer/.test(blob)) return true;
      if (key.includes('medical') && /medicine|medical|health/.test(blob)) return true;
      if (key.includes('engineer') && blob.includes('engineering')) return true;
      if (key.includes('business') && blob.includes('business')) return true;
      return false;
    });
  });
}

function avgMarks() {
  const vals = Object.values(state.academic.marks || {}).map(markPercent).filter((n) => n != null && !Number.isNaN(n));
  if (!vals.length) return 0;
  return Math.round(vals.reduce((a, b) => a + b, 0) / vals.length);
}
